# Data Cleaning

## Project: Retail Sales Data Analytics

This notebook focuses on preparing the dataset for analysis by addressing data quality issues and ensuring consistency.

The key steps include:
- Handling missing values
- Fixing data types
- Removing duplicates
- Preparing a clean dataset for further analysis

In [1]:
import pandas as pd

df = pd.read_csv("../data/raw/Sample - Superstore.csv", encoding="latin1")
df.head()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,6/12/2016,6/16/2016,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164


In [2]:
df.isnull().sum()

Row ID           0
Order ID         0
Order Date       0
Ship Date        0
Ship Mode        0
Customer ID      0
Customer Name    0
Segment          0
Country          0
City             0
State            0
Postal Code      0
Region           0
Product ID       0
Category         0
Sub-Category     0
Product Name     0
Sales            0
Quantity         0
Discount         0
Profit           0
dtype: int64

## Missing Values Handling

The dataset was checked for missing values across all columns.

No missing values were found, indicating that the dataset is complete and does not require imputation or removal of records at this stage.

This ensures that subsequent analysis will not be affected by incomplete data.

In [3]:
df['Order Date'] = pd.to_datetime(df['Order Date'])
df['Ship Date'] = pd.to_datetime(df['Ship Date'])

df[['Order Date', 'Ship Date']].dtypes

Order Date    datetime64[ns]
Ship Date     datetime64[ns]
dtype: object

## Data Type Conversion

The `Order Date` and `Ship Date` columns were originally stored as object (string) data types.

These columns were converted to datetime format to enable time based analysis, such as trend analysis, time series evaluation, and feature extraction (e.g., year, month).

This ensures that date related operations can be performed accurately in later stages of the project.

In [4]:
df.duplicated().sum()

0

## Duplicate Records Handling

The dataset was checked for duplicate records using the `duplicated()` function.

No duplicate rows were found in the dataset, indicating that each transaction is unique.

Therefore, no records were removed at this stage.

In [5]:
(df['Sales'] < 0).sum()

0

## Sales Consistency Check

The dataset was checked for negative values in the Sales column.

No negative values were found, indicating that all recorded sales amounts are valid and consistent with expected business rules.

This confirms that the Sales data does not require correction at this stage.

In [6]:
(df['Profit'] < 0).sum()

1871

## Profit Consistency Check

The dataset was checked for negative values in the Profit column.

A total of 1871 transactions were found to have negative profit values.

Unlike Sales, negative Profit values are valid and indicate loss-making transactions. These may result from factors such as high discounts, low pricing, or high costs.

These records were retained, as they provide important insights into business performance and will be further analyzed in later stages of the project.

In [7]:
df.to_csv("../data/processed/cleaned_superstore.csv", index=False)

## Summary

This stage prepared the dataset for analysis by ensuring data quality, consistency, and feature availability.